File: java/sodium/src/sodium/EventLoop.java
package sodium;

import java.util.List;

public class EventLoop<A> extends Event<A> {
    private Event<A> ea_out;

    public EventLoop()
    {
    }

    // TO DO: Copy & paste from EventSink. Can we improve this?
    private void send(Transaction trans, A a) {
        if (firings.isEmpty())
            trans.last(new Runnable() {
            	public void run() { firings.clear(); }
            });
        firings.add(a);

        @SuppressWarnings("unchecked")
		List<TransactionHandler<A>> listeners = (List<TransactionHandler<A>>)this.listeners.clone();
    	for (TransactionHandler<A> action : listeners) {
    		try {
                action.run(trans, a);
    		}
    		catch (Throwable t) {
    		    t.printStackTrace();
    		}
    	}
    }

    public void loop(Event<A> ea_out)
    {
        if (this.ea_out != null)
            throw new RuntimeException("EventLoop looped more than once");
        this.ea_out = ea_out;
        final EventLoop<A> me = this;
        addCleanup(ea_out.listen_(this.node, new TransactionHandler<A>() {
            public void run(Transaction trans, A a) {
                me.send(trans, a);
            }
        }));
    }
}

File: c#/Sodium/src/Sodium/EventLoop.cs
using System;
using System.Collections.Generic;

namespace Sodium
{

  public class EventLoop<TA> : Event<TA>
  {
    Event<TA> eaOut;

    public EventLoop()
    {
      if (Transaction.GetCurrentTransaction() == null)
        throw new Exception("EventLoop/BehaviorLoop must be used within an explicit transaction");

      SampleNow = () =>
      {
        if (eaOut == null)
          throw new Exception("EventLoop sampled before it was looped");
        return eaOut.SampleNow();
      };
    }


    // TO DO: Copy & paste from EventSink. Can we improve this?
    void Send(Transaction trans, TA a)
    {
      if (Firings.IsEmpty())
        trans.Last(
          new Runnable
          {
            Run = () => Firings.Clear()
          });
      Firings.Add(a);

      var listeners = (List<TransactionHandler<TA>>)Listeners.Clone();
      foreach (TransactionHandler<TA> action in listeners)
      {
        try
        {
          action.Run(trans, a);
        }
        catch (Exception e)
        {
          Console.WriteLine(e);
        }
      }
    }

    public void Loop(Event<TA> eventOut)
    {
      if (this.eaOut != null)
        throw new Exception("EventLoop looped more than once");
      this.eaOut = eventOut;
      EventLoop<TA> me = this;
      AddCleanup(
        eventOut.Listen_(
          Node,
          new TransactionHandler<TA>
          {
            Run = (trans, a) => me.Send(trans, a)
          }));
    }
  }

}